playwright/fixtures/common.ts

```diff
@@ -0,0 +1,100 @@
+import { _electron as electron } from "playwright";
+import { test as base, expect } from "@playwright/test";
+import * as fs from "fs";
+import * as path from "path";
+import * as crypto from 'crypto';
+
+export function generateUUID(): string {
+  return crypto.randomBytes(16).toString('hex');
+}
+
+type TestFixtures = {
+  lang: String;
+  theme: String;
+  userdata: String;
+  env: Object;
+  page: any;
+};
+
+const test = base.extend<TestFixtures>({
+  userdata: undefined,
+  env: undefined,
+  lang: "en-US",
+  theme: "light",
+  page: async ({ lang, theme, userdata, env }, use) => {
+    // create userdata path
+    const userDataPathKey = generateUUID();
+    const userDataPath = path.join(__dirname, "../artifacts/userdata", userDataPathKey);
+    fs.mkdirSync(userDataPath, { recursive: true });
+
+    if (userdata) {
+      const userDataFile = path.resolve("playwright/userdata/", `${userdata}.json`);
+      fs.copyFileSync(userDataFile, `${userDataPath}/app.json`);
+    }
+
+    // default environment variables
+    env = Object.assign(
+      {
+        MOCK: true,
+        HIDE_DEBUG_MOCK: true,
+        CI: process.env.CI || undefined,
+        SPECTRON_RUN: true,
+        // SYNC_ALL_INTERVAL: 86400000,
+        // SYNC_BOOT_DELAY: 16,
+      },
+      env,
+    );
+
+    // launch app
+    const viewport = { width: 1024, height: 768 };
+
+    const electronApp = await electron.launch({
+      args: [
+        "./.webpack/main.bundle.js",
+        `--user-data-dir=${userDataPath}`,
+        `--window-size=${viewport.width},${viewport.height}`,
+      ],
+      recordVideo: {
+        dir: "playwright/artifacts/videos/",
+        size: viewport,
+      },
+      env,
+      colorScheme: theme,
+      locale: lang,
+    });
+
+    // app is ready
+    const page = await electronApp.firstWindow();
+
+    // start coverage
+    const istanbulCLIOutput = path.join('playwright/artifacts/.nyc_output');
+
+    await page.addInitScript(() =>
+    window.addEventListener('beforeunload', () =>
+      (window as any).collectIstanbulCoverage(JSON.stringify((window as any).__coverage__))
+      ),
+    );
+    await fs.promises.mkdir(istanbulCLIOutput, { recursive: true });
+    await page.exposeFunction('collectIstanbulCoverage', (coverageJSON: string) => {
+      if (coverageJSON)
```

Comment:
Could we not move this to the start of the code coverage section so that we don't run it if it's not needed? We seem to be doing extra things that may not be needed

Reply:
Agreed, I'm waiting for *Plawright + CI* brainstorm to improve this, so when we develop tests locally, we won't do those extra things.